feat(docs): add hash custom instructions and rearange guides (#1400)
flare-foundation/developer-hub: docs/smart-accounts/5-custom-instruction-comparison.mdx

---
sidebar_position: 1
slug: custom-instruction-comparison
title: Custom Instruction Comparison
authors: [nikerzetic]
description: Comparing the raw (0xFF) and hash-based (0xFE) custom instruction flows for Flare Smart Accounts.
tags: [intermediate, ethereum, flare-smart-accounts]
keywords:
  [
    flare-fdc,
    ethereum,
    flare-smart-accounts,
    evm,
    flare-network,
    account-abstraction,
  ]
---

> Flare Smart Accounts expose two custom instruction memo opcodes that ultimately execute the same `PackedUserOperation` against a personal account:

Reviewer: ```suggestion
Flare Smart Accounts expose two custom instruction memo opcodes that ultimately execute the same `PackedUserOperation` in the personal account scope:
```

- [**Custom Instruction**](/smart-accounts/custom-instruction) - opcode `0xFE`.
  The XRPL memo carries only `keccak256(userOp)` in fixed 42 bytes; an off-chain executor delivers the ABI-encoded custom instruction (`userOp`) via `executeDirectMintingWithData`.
- [**Raw Custom Instruction**](/smart-accounts/raw-custom-instruction) - opcode `0xFF`.
  The XRPL memo contains the ABI-encoded `PackedUserOperation` in full.

Both flows are validated on-chain against the same `(sender, nonce)` rules and emit the same [`UserOperationExecuted`](/smart-accounts/reference/IMasterAccountController#useroperationexecuted) event.
The difference is purely in how the user-operation bytes travel to Flare, and which actor performs which step.

The hash-based flow was added because the XRPL `Payment` memo is capped at `1024` bytes - small enough that non-trivial user operations either had to be split across multiple XRPL payments (each paying its own minting and executor fees) or routed through purpose-built shim contracts on Flare that compressed the call into something memo-sized.

> The 0xFE memo is a constant 42 bytes regardless of the user operation size, which removes the need for both workarounds.

Reviewer: ```suggestion
The `0xFE` memo is a constant 42 bytes regardless of the user's operation size, removing the need for both workarounds.
```

> ## Side-by-side

Reviewer: ```suggestion
## Comparison Table
```

> | Dimension                          | Custom Instruction (`0xFE`)                                                                   | Raw Custom Instruction (`0xFF`)                                                           |

Reviewer: Let's add link to both instruction type page

| AssetManager entry point           | `executeDirectMintingWithData(proof, data)`                                                   | `executeDirectMinting(proof)`                                                             |
| `msg.value` on the inner call      | Whatever the executor attaches; forwarded to `executeUserOp`                                  | Whatever the relayer attaches; forwarded to `executeUserOp`                               |
| Confirmation                       | Same event lives in the executor's transaction receipt                                        | Watch for `UserOperationExecuted` after the `MasterAccountController` dispatches the memo |
| Replay protection                  | Same as raw - plus the on-chain hash check pins `_data` to the memo's commitment              | Personal account nonce + `usedTransactionIds`                                             |
| Failure modes specific to the flow | `CustomInstructionHashMismatch` if `keccak256(_data)` differs from the commitment             | None beyond shared ones                                                                   |

> ## When to use which

Reviewer: I think this header is not needed, there

rename `Comparison Table` to `Comparison`

- You want the call payload to remain off the public XRPL ledger.
- You already operate an executor that observes XRPL payments and submits Flare transactions - the custom instruction lets that executor batch deliveries or rate-limit submissions without splitting user operations into multiple XRPL payments.
- You need fee predictability: the XRPL memo is always 42 bytes, so the XRPL fee is constant regardless of how complex the user operation is.

**Use the raw custom instruction (`0xFF`) when**

- The call batch fits comfortably inside the XRPL ~1024-byte memo (most single-call user operations and small batches do).
- You do not need to hide the call payload from public XRPL observers.
- You do not want to operate or coordinate with an executor service.
- You want the simplest possible on-chain integration: a single FAssets transaction `executeDirectMinting(proof)` carries the whole thing.

In short, the custom instruction trades a small amount of off-chain coordination (and one extra on-chain proof-binding check) for a fixed memo size and private payloads.
The raw custom instruction trades memo bandwidth for end-to-end simplicity.

> ## What stays the same

Reviewer: ```suggestion
## Similarities
```
